### DISCUSSION TITLE
Failing to run Traffic Generation simulation due to unsupported system calls

### DISCUSSION THREAD

**Discussion by afressancourt** [shadow/shadow#2299] [Q&A] [answered]
``Hello,

I am sorry to bother you, but following the patch you submitted last week, I have tried to run the TGen simulation presented in teh documentation, and I encounter a problem doing so.

Indeed, when I try to run the simulation, it seems to hang and the log show the following lines repeatedly:
```
00:00:00.543957 [703233:shadow-worker] 00:00:01.000000000 [WARN] [server:11.0.0.11] [syscall_handler.c:527] [syscallhandler_make_syscall] Detected unsupported syscall 435 called from thread 1000 in process server.tgen.1000 on host server
00:00:00.543958 [703233:shadow-worker] 00:00:01.000000000 [ERROR] [server:11.0.0.11] [syscall_handler.c:531] [syscallhandler_make_syscall] Returning error 38 (ENOSYS) for unsupported syscall 435, which may result in unusual behavior
00:00:00.552299 [703233:shadow-worker] 00:00:01.000000000 [WARN] [server:11.0.0.11] [syscall_handler.c:527] [syscallhandler_make_syscall] Detected unsupported syscall 204 called from thread 1000 in process server.tgen.1000 on host server
00:00:00.552300 [703233:shadow-worker] 00:00:01.000000000 [ERROR] [server:11.0.0.11] [syscall_handler.c:531] [syscallhandler_make_syscall] Returning error 38 (ENOSYS) for unsupported syscall 204, which may result in unusual behavior
00:00:00.554714 [703233:shadow-worker] 00:00:01.000000000 [WARN] [server:11.0.0.11] [syscall_handler.c:527] [syscallhandler_make_syscall] Detected unsupported syscall 237 called from thread 1001 in process server.tgen.1000 on host server
00:00:00.554716 [703233:shadow-worker] 00:00:01.000000000 [ERROR] [server:11.0.0.11] [syscall_handler.c:531] [syscallhandler_make_syscall] Returning error 38 (ENOSYS) for unsupported syscall 237, which may result in unusual behavior
```
(TL;DR: Syscalls 435, 204 and 237 are not supported)

I compiled and installed TGen without obvious problems for what I can determine.

As in my earlier posts, I am running on a Ubuntu 22.04 LTS, with kernel:
```
Linux <REDACTED> 5.15.0-39-generic #42-Ubuntu SMP Thu Jun 9 23:42:32 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

Is this issue related to the issue with trampoline symbol injection? 

Thanks a lot in advance for your answer,

Antoine

**Comment by sporksmith** [comment]
@afressancourt thank you for the report, and sorry you keep running into these stumbling blocks! Identifying these issues is valuable to us, and we're more than happy to dig into them. I hope it's not souring you on using shadow though; if you'd rather have a smoother experience, you'd probably have better luck with a slightly older distro.

> 00:00:00.543957 [703233:shadow-worker] 00:00:01.000000000 [WARN] [server:11.0.0.11] [syscall_handler.c:527] [syscallhandler_make_syscall] Detected unsupported syscall 435 called from thread 1000 in process server.tgen.1000 on host server
```
$ grep 'NR.*\b435' /usr/include/x86_64-linux-gnu/asm/unistd_64.h
#define __NR_clone3 435
```

This is `clone3`, which is a new version of `clone`. This is almost certainly what's breaking the simulation. I *think* it shouldn't be too difficult to implement.

This is the first time I've seen anything trying to use it. I wonder why our new CI tests  (which use ubuntu 22.04 both in the base system and container) aren't hitting this...


> 00:00:00.552299 [703233:shadow-worker] 00:00:01.000000000 [WARN] [server:11.0.0.11] [syscall_handler.c:527] [syscallhandler_make_syscall] Detected unsupported syscall 204 called from thread 1000 in process server.tgen.1000 on host server

This is `sched_getaffinity`. It's on the roadmap to implement, but typically doesn't actually break anything in the meantime. https://github.com/shadow/shadow/issues/2101

> 00:00:00.554714 [703233:shadow-worker] 00:00:01.000000000 [WARN] [server:11.0.0.11] [syscall_handler.c:527] [syscallhandler_make_syscall] Detected unsupported syscall 237 called from thread 1001 in process server.tgen.1000 on host server

This is `mbind`. As with `sched_getaffinity` we ought to implement it, but it's probably not breaking things in the meantime.

I'll open an issue for clone3 and try to take a look at it today.

**Reply by sporksmith** [reply]
Opened https://github.com/shadow/shadow/issues/2300 to track clone3

**Comment by sporksmith** [comment]
Actually, `clone3` might be a red herring; I wouldn't be surprised if glibc had a fallback to use `clone` when the `clone3` call fails, which would explain why we haven't seen related failures.

Could you attach shadow's output, and the generated `shadow.data` directory? Preferably using a debug build of shadow, and with shadow flags `--strace-logging-mode=standard` and `--log-level=trace`.

**Reply by sporksmith** [reply]
(you'll probably need to compress everything into a zip or tar.gz archive to pass github's attachment size and type restrictions)

**Reply by afressancourt** [reply]
No worries, it doesn't prevent me from trying to run Shadow. By the way, I have run the minimal Tor setup and it works fine. As this is ultimately the scenario I was willing to run, I am really pleased with the fantastic work you have been doing.

I will rebuild and try to push the logs I have for you to have debug information as soon as possible.

**Reply by afressancourt** [reply]
(If you could just point me to the command line I need to use, I have tried to give the compilation flags to the setup script and it doesn't work :-/)

**Reply by sporksmith** [reply]
To compile and install a debug build: `./setup build --debug && ./setup install`

The other flags I mentioned should be passed to shadow itself. So something like:
`shadow --strace-logging-mode=standard --log-level=trace mysimconfig.yml > shadow.log`

**Reply by afressancourt** [reply]
Hello,
 
I will have trouble uploading the log files because of some restrictions on my side, but in the meantime I managed to dig into the log to help a bit.

After a first look at the log files, I realized that what was happening is that the server is the only node that has actually done something, and none of the 10 clients have done a thing (there is not any trace in the `shadow.data` directory for those clients).

In the `shadow.log` file, I spotted the following line:
```
00:00:00.145903 [757311:shadow-worker] 00:00:01.000000000 [TRACE] [server:11.0.0.11] [thread_preload.c:246] [_threadpreload_waitForNextEvent] received shim_event 3
```
That comes regularly.

If I look in the `shadow.data/hosts/server/server.tgen.1000.shimlog` file I can see a loop of the three following log lines:

```
00:00:01.475372 [00:00:01.000000000] [shd-shim] [trace] [shim_syscall.c:267] [shim_syscallv] Handled syscall 24 from the shim; we avoided inter-process overhead.
00:00:01.475378 [00:00:01.000000000] [shd-shim] [trace] [shim_rdtsc.c:25] [_shim_rdtsc_handle_sigsegv] Trapped sigsegv
00:00:01.475382 [00:00:01.000000000] [shd-shim] [trace] [shim_rdtsc.c:56] [_shim_rdtsc_handle_sigsegv] Emulating rdtsc
```

In the `shadow.log` file, after a while it mentions that the server process is blocking and running without interruption. Then, I can see the result of my forced interruption.

**Reply by sporksmith** [reply]
Thanks!

This looks like #1788 (which itself is the tgen-specific instance of #1792).

Looking at the [tgen example in our docs](https://shadow.github.io/docs/guide/getting_started_tgen.html#configuring-shadow), I see it doesn't have either of the known solutions. It should be either setting the environment variable `OPENBLAS_NUM_THREADS=1` in the tgen processes (a tgen-specific way to force it to avoid busy-waiting), or enabling [`model_unblocked_syscall_latency`](https://shadow.github.io/docs/guide/shadow_config_spec.html#generalmodel_unblocked_syscall_latency) (our general solution for getting out of busy-wait loops).

@afressancourt could you try one of those? The easiest way to try it is to add `--model-unblocked-syscall-latency=true`  at the command-line.

**Reply by sporksmith** [reply]
Btw for more about these issues, the user-facing documentation might be easier to read through than the github issues:

* [busy loops](https://shadow.github.io/docs/guide/limitations.html#busy-loops)
* [libopenblas compatibility notes](https://shadow.github.io/docs/guide/compatibility_notes.html#libopenblas)

**Reply by afressancourt** [reply]
Hello,

Thanks a lot for your answer ! 

I had to set the environment variable in the YAML file for the scenario to execute successfully, otherwise there is still a deadlock on the server node when it is started.

When I only use the command line option, I see the following lines repeatedly in `shadow.data/hosts/server/server.tgen.1000.shimlog` :
```
00:00:14.428813 [00:00:01.001572000] [shd-shim] [trace] [shim_syscall.c:128] [_shim_emulated_syscall_event] got response of type 4 on 0x7ffff23ee808
00:00:14.428822 [00:00:01.001572000] [shd-shim] [trace] [shim_syscall.c:267] [shim_syscallv] Handled syscall 24 from the shim; we avoided inter-process overhead.
00:00:14.428830 [00:00:01.001572000] [shd-shim] [trace] [shim_rdtsc.c:25] [_shim_rdtsc_handle_sigsegv] Trapped sigsegv
00:00:14.428834 [00:00:01.001572000] [shd-shim] [trace] [shim_rdtsc.c:56] [_shim_rdtsc_handle_sigsegv] Emulating rdtsc
00:00:14.428840 [00:00:01.001572000] [shd-shim] [trace] [shim_sys.c:157] [shim_sys_handle_syscall_locally] unappliedCpuLatency=1000 maxUnappliedCpuLatency=1000
00:00:14.428845 [00:00:01.001572000] [shd-shim] [trace] [shim_syscall.c:267] [shim_syscallv] Handled syscall 24 from the shim; we avoided inter-process overhead.
00:00:14.428851 [00:00:01.001572000] [shd-shim] [trace] [shim_rdtsc.c:25] [_shim_rdtsc_handle_sigsegv] Trapped sigsegv
00:00:14.428856 [00:00:01.001572000] [shd-shim] [trace] [shim_rdtsc.c:56] [_shim_rdtsc_handle_sigsegv] Emulating rdtsc
00:00:14.428861 [00:00:01.001572000] [shd-shim] [trace] [shim_sys.c:157] [shim_sys_handle_syscall_locally] unappliedCpuLatency=2000 maxUnappliedCpuLatency=1000
00:00:14.428866 [00:00:01.001572000] [shd-shim] [trace] [shim_sys.c:181] [shim_sys_handle_syscall_locally] Reached maxUnappliedCpuLatency. Yielding. (2000 ns past max)
00:00:14.428870 [00:00:01.001572000] [shd-shim] [trace] [shim_syscall.c:271] [shim_syscallv] Making syscall 1005 indirectly; we ask shadow to handle it using the shmem IPC channel.
00:00:14.428875 [00:00:01.001572000] [shd-shim] [trace] [shim_syscall.c:115] [_shim_emulated_syscall_event] sending syscall 1005 event on 0x7ffff23ee808
00:00:14.428880 [00:00:01.001572000] [shd-shim] [trace] [shim_syscall.c:277] [shim_syscallv] Making syscall 24 directly; we expect ptrace or seccomp will interpose it, or it will be handled natively by the kernel.
00:00:14.428885 [00:00:01.001572000] [shd-shim] [trace] [shim_syscall.c:125] [_shim_emulated_syscall_event] waiting for event on 0x7ffff23ee808
00:00:14.428890 [00:00:01.001572000] [shd-shim] [trace] [shim_syscall.c:277] [shim_syscallv] Making syscall 202 directly; we expect ptrace or seccomp will interpose it, or it will be handled natively by the kernel.
```

**Reply by sporksmith** [reply]
Thanks for the follow-up!

Looks like I need to add a test with an rdtsc-only busy-loop, and update the tgen example in the docs.

**Comment by sporksmith** [comment, marked as answer]
To close this out:

Ultimately the failure was due to a "busy loop" in libopenblas, which tgen depends on via libigraph. We'd already documented this issue and [workarounds](https://shadow.github.io/docs/guide/compatibility_notes.html#libopenblas), but had forgotten to include the workarounds in one of the examples. Additionally, one of the workarounds (`--model-unblocked-syscall-latency`) didn't work for some versions of libopenblas, which have a busy-loop that only uses rdtsc to measure time.

Both of these issues are now fixed:

* The example now includes the workaround: https://github.com/shadow/shadow/pull/2312
* We now handle rdtsc-only busy loops: https://github.com/shadow/shadow/pull/2314
